### PR TITLE
Magnolia hardfork

### PR DESCRIPTION
**Pull request by JakubFornadel**
## Purpose

Magnolia hardfork contains 3 major changes:
- validator deletion fix -> delete validator only after he confirms his undelegation
- slashing contract -> when someone double votes, he is jailed for some time
- rewards distribution change

evm pr: https://github.com/Taraxa-project/taraxa-evm/pull/184

TODO:
- I did set jail_time to 7 days - we should agree on some value

### CODE REVIEW

File: libraries/cli/include/cli/config_jsons/default/default_genesis.json [hunk 1]
```diff
@@ -60,6 +60,11 @@
     "2cd4da7d3b345e022ca7e997c2bb3276a4d3d2e9": "0x1027e72f1f12813088000000",
     "7e4aa664f71de4e9d0b4a6473d796372639bdcde": "0x1027e72f1f12813088000000"
   },
+  "hardforks": {
+    "rewards_distribution_frequency": {
```
Comment (JakubFornadel) [inline libraries/cli/include/cli/config_jsons/default/default_genesis.json:64]: @kstdl do we need this frequency ? it is not used anywhere

File: libraries/core_libs/consensus/src/rewards/block_stats.cpp [hunk 2]
```diff
@@ -6,33 +6,40 @@
...
-    return {};
+bool BlockStats::addTransaction(const trx_hash_t& trx_hash, const addr_t& validator) {
+  auto itr = fee_by_trx_hash_.find(trx_hash);
+  if (itr == fee_by_trx_hash_.end()) {
+    // No fee record found for transaction means that transaction was processed before
+    return false;
   }
 
-  return {found_tx->second};
+  validators_stats_[validator].fees_rewards_ += itr->second;
```
Comment (JakubFornadel) [inline libraries/core_libs/consensus/src/rewards/block_stats.cpp:38]: @kstdl I saw in evm that we no longer use tx fees in DistributeRewards function, what happens with txs fees now ?

File: libraries/cli/include/cli/config_jsons/default/default_genesis.json (hunk 1 above)
Comment (kstdl) [reply libraries/cli/include/cli/config_jsons/default/default_genesis.json]: https://github.com/Taraxa-project/taraxa-node/pull/2545/files#diff-e2f034014fd538f8b11e72f17b1376627107ac1247e49443449e5319bcbf57b1R30
It is used here

File: libraries/core_libs/consensus/src/rewards/block_stats.cpp (hunk 2 above)
Comment (kstdl) [reply libraries/core_libs/consensus/src/rewards/block_stats.cpp:38]: https://github.com/Taraxa-project/taraxa-evm/pull/184/files#diff-926216bbe8192eb41a788c8d0b3f6a4a37932ac63db9613e1ba354e5cfb787c9R705-R713

File: libraries/cli/include/cli/config_jsons/default/default_genesis.json [hunk 3]
```diff
@@ -54,7 +54,10 @@
         },
         "vrf_key": "f8d5c00ce9fa3058341e051b36a1e6ccf69df81fb865568b2bf1507d085691e2"
       }
-    ]
+    ],
+    "slashing": {
+      "jail_time": 163459
+    }
```
[the remark this answers is not in the document]
Comment (JakubFornadel) [reply libraries/cli/include/cli/config_jsons/default/default_genesis.json:60]: yeah thats the thing I wanted to discuss and forgot about it. I will put it there

**Review by mfrankovi — APPROVED** [review]

**Review by kstdl — APPROVED** [review]

**Pull request merged**